src/node.rs

```diff
@@ -0,0 +1,1134 @@
+        let mut new_node = self.clone();
+        let idx = new_node.child_ptr_indexes.get(key as usize).unwrap();
+        new_node.children.set(*idx as usize, node);
+        new_node.update_ts_to_max_child_ts();
+
+        new_node
+    }
+
+    fn add_child(&self, key: u8, node: N) -> Self {
+        let mut new_node = self.clone();
+
+        // Update the timestamp if the new child has a greater timestamp
+        new_node.update_if_newer(node.ts());
+
+        new_node.insert_child(key, Arc::new(node));
+        new_node
+    }
+
+    fn delete_child(&self, key: u8) -> Self {
+        let pos = self.child_ptr_indexes.get(key as usize).unwrap();
+        let mut new_node = self.clone();
+        new_node.child_ptr_indexes.erase(key as usize);
+        new_node.children.erase(*pos as usize);
+        new_node.num_children -= 1;
+
+        new_node.update_ts_to_max_child_ts();
+        new_node
+    }
+
+    fn find_child(&self, key: u8) -> Option<&Arc<N>> {
+        let idx = self.child_ptr_indexes.get(key as usize)?;
+        let child = self.children.get(*idx as usize)?;
+        Some(child)
+    }
+
+    fn num_children(&self) -> usize {
+        self.num_children as usize
+    }
+
+    #[inline]
+    fn size(&self) -> usize {
+        48
+    }
+}
+
+impl<P: KeyTrait + Clone, N: Timestamp> Timestamp for Node48<P, N> {
+    fn ts(&self) -> u64 {
+        self.ts
+    }
+}
+
+// Source: https://www.the-paper-trail.org/post/art-paper-notes/
+//
+// Node256: It is the traditional trie node, used when a node has
+// between 49 and 256 children. Looking up child pointers is obviously
+// very efficient - the most efficient of all the node types - and when
+// occupancy is at least 49 children the wasted space is less significant.
+//
+// A Node256 is a 256-entry array of pointers to children. The pointers are stored in
+// a Vector Array, which is a Vector of length WIDTH (256) that stores the pointers.
+pub struct Node256<P: KeyTrait + Clone, N: Timestamp> {
+    pub(crate) prefix: P, // Prefix associated with the node
+    pub(crate) ts: u64,   // Timestamp for node256
+
+    children: Box<SparseArray<Arc<N>, 256>>,
+    num_children: usize,
+}
+
+impl<P: KeyTrait + Clone, N: Timestamp> Node256<P, N> {
+    pub fn new(prefix: P) -> Self {
+        Self {
+            prefix,
+            ts: 0,
+            children: Box::new(SparseArray::new()),
+            num_children: 0,
+        }
+    }
+
+    pub fn shrink(&self) -> Node48<P, N> {
+        let mut indexed = Node48::new(self.prefix.clone());
+        let keys: Vec<usize> = self.children.iter_keys().collect();
+        for key in keys {
+            let child = self.children.get(key).unwrap().clone();
+            indexed.insert_child(key as u8, child);
+        }
+        indexed.update_ts();
+        indexed
+    }
+
+    #[inline]
+    fn insert_child(&mut self, key: u8, node: Arc<N>) {
+        self.children.set(key as usize, node);
+        self.num_children += 1;
```

Comment:
I'm curious if there is any chance the child already exists and you don't need to increment `num_children`?

Reply:
Totally makes sense. This is an internal method being used by add_child which is called by the Trie post checks of whether the key exists or not at that position.

Reply:
Maybe a `debug_assert`, then :wink: